### PR TITLE
[Merged by Bors] - tortoise: refactor trtl.Persist to use its own store and do the task more efficiently

### PR DESCRIPTION
**Pull request by dshulyak**
## Motivation
My main motivation was to simplify encoding in turtle.Persist. turtle.Persist also shows up in profiles (both in cpu and allocations) and i thought that it will be more efficient without encoding whole thing at once.

```
BenchmarkStatePersist
BenchmarkStatePersist/New
BenchmarkStatePersist/New-16         	    4899	    252503 ns/op	  140557 B/op	    2625 allocs/op
BenchmarkStatePersist/Repeat
BenchmarkStatePersist/Repeat-16      	   52162	     22840 ns/op	     496 B/op	      16 allocs/op
BenchmarkStatePersist/Old
BenchmarkStatePersist/Old-16         	    2386	    461755 ns/op	  148863 B/op	    5432 allocs/op

```

BenchmarkStatePersist/Repeat is using diff mode. It still has to loop through every diff, we can improve it by writing to the database in HandleIncomingLayer directly but it will require a bit more changes.

related to: https://github.com/spacemeshos/go-spacemesh/issues/2545
## Changes
- added state to the turtle instead of relying on mesh.TORTOISE key, which seemed quite hacky to me
- tested state recovery with testing/quick, also added simple benchmark for new and old Persist version
- simplified initialized
- added support for diff mode, that will persist only non-flushed records from collections and layers

## Test Plan
unit tests

## TODO
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by dshulyak** [comment]
this can wait until self-healing is merged, i will fix conflicts afterwards

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
@lrettig i think we will want to merge this change into/before self-healing. if i read profiles correctly node will stop crashing. unless there are other ideas

File: tortoise/state.go
Comment (antonlerner) [inline tortoise/state.go:73]: mark as flushed after `batch.Put` succeeds

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

File: tortoise/util.go
Comment (lrettig) [inline tortoise/util.go]: We use uint64 for this in healing branch. see #2629

File: tortoise/verifying_tortoise.go
Comment (lrettig) [inline tortoise/verifying_tortoise.go]: should be uint64, weight can never be negative

File: tortoise/verifying_tortoise.go
Comment (lrettig) [inline tortoise/verifying_tortoise.go]: this should be `true` shouldn't it?

File: tortoise/verifying_tortoise.go
Comment (lrettig) [inline tortoise/verifying_tortoise.go]: why perform eviction inside persist (rather than when verifying a layer, as we do now)?

File: tortoise/verifying_algorithm.go
Comment (lrettig) [inline tortoise/verifying_algorithm.go]:
```suggestion
			cfg.Log.With().Panic("can't recover turtle state", log.Err(err))
```

File: tortoise/verifying_tortoise.go
Comment (lrettig) [inline tortoise/verifying_tortoise.go:162]: don't you still need to initialize the `Opinion` map?

File: tortoise/verifying_tortoise.go
Comment (lrettig) [inline tortoise/verifying_tortoise.go]: should be `true`

File: tortoise/state.go
Comment (lrettig) [inline tortoise/state.go]:
```suggestion
					s.log.With().Panic("can't encode vec", log.Err(err))
```

**Review by lrettig — COMMENTED** [review]
I'm not sure how I feel about the custom encoder for tortoise. Let's discuss with @antonlerner

File: tortoise/state.go
Comment (dshulyak) [reply tortoise/state.go:73]: it doesn't matter. actual write will happen in batch.Write

File: tortoise/verifying_tortoise.go
Comment (dshulyak) [reply tortoise/verifying_tortoise.go:162]: Opinion is a map now

File: tortoise/verifying_tortoise.go
Comment (dshulyak) [reply tortoise/verifying_tortoise.go]: i kept it here because beacause HandleIncomingLayer wasn't exposing error

**Comment by dshulyak** [comment]
> I'm not sure how I feel about the custom encoder for tortoise. 

what is custom encoder? writing state the way it is implemented now is not an option tbh

as i mentioned already, encoding 200mb of state requires a lot of memory allocated at once by codec. also every write to a leveldb (or in fact any lsm based storage) is not an inplace update like with sqlite or postgres. it appends data and later it needs to compact this state with lower levels, which will use a lot of memory and will block other writes. so current solution is simply not something that can be used in principal 

one node in the devnet stores state for 1000 layers and is still far from crashing. previously it was crashing once it stored state for 150 layers.

File: tortoise/verifying_tortoise.go
Comment (dshulyak) [reply tortoise/verifying_tortoise.go]: true is for flushed

File: tortoise/verifying_tortoise.go
Comment (dshulyak) [reply tortoise/verifying_tortoise.go]: true is for flushed

File: tortoise/util.go
Comment (dshulyak) [reply tortoise/util.go]: can it be limited to uint32? will save some inuse memory. given that we will be storing millions of votes in mainnet

Comment (dshulyak) [reply tortoise/util.go]: based on last observation, opinion layers 18-1057. total 6173753. evict from 28. last 1057. verified 38. reducing it to uint32 will save ~50mb. so maybe won't be that important

**Comment by antonlerner** [comment]
>one node in the devnet stores state for 1000 layers and is still far from crashing. previously it was crashing once it stored state for 150 layers.

why is the state so large? this is probably the root cause for this large memory consumption

**Comment by dshulyak** [comment]
> why is the state so large? this is probably the root cause for this large memory consumption

are you asking about 1000 or 150? 150 we had in the last devnet that crashed, i think @lrettig can explain what size we need. i saw a comment that in mainnet we will want to store for 1000. 

anyway this pattern is still super inefficient

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
failed with k8s error in latenodes:
```
E           HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"DaemonSet in version \"v1\" cannot be handled as a DaemonSet: v1.DaemonSet.Spec: v1.DaemonSetSpec.Template: v1.PodTemplateSpec.Spec: v1.PodSpec.NodeSelector: ReadString: expects \" or n, but found 5, error found in #10 byte of ...|espace\": 59409}, \"to|..., bigger context ...|eriodSeconds\": 10, \"nodeSelector\": {\"namespace\": 59409}, \"tolerations\": [{\"key\": \"namespace\", \"opera|...","reason":"BadRequest","code":400}
```

bors try

File: tortoise/util.go
Comment (lrettig) [reply tortoise/util.go]: no, it needs to encode voting weights, and voting weights are uint64. this would be a very deep change to varpost.

File: tortoise/verifying_tortoise.go
Comment (lrettig) [reply tortoise/verifying_tortoise.go:162]: exactly. don't you need to initialize it with `make`?

File: tortoise/verifying_tortoise.go
Comment (lrettig) [reply tortoise/verifying_tortoise.go]: I don't understand. this line (and the other, similar one) very clearly looks like it's marking the block with id `id` as not good (`false`). am I missing something?

File: tortoise/verifying_tortoise.go
Comment (dshulyak) [reply tortoise/verifying_tortoise.go:162]: but Opinion{} and make(Opinion) will do the same thing for map, unless you want to specify length

File: tortoise/verifying_tortoise.go
Comment (dshulyak) [reply tortoise/verifying_tortoise.go]: so if block is present in this map it means that the block is good. if the value is set to true it means that it was flushed to the disk, and doesn't need to be flushed again

**Comment by dshulyak** [comment]
last stats from a devnet using this branch + self healing: opinion layers 134-338. total 7358437. last evicted 236. last 338. verified 337. so the opinions size seems to be bounded by ~2*windowSize. it consumes around ~700mb to store this data in memory. but the good thing is that network is not even close to crashing due to memory usage.

@lrettig can you clarify please if we need to support windowSize = 1000 in mainnet? it basically will require each node to use 7gb just to store votes, so we will have to figure out something in order to support that.

**Comment by lrettig** [comment]
> @lrettig can you clarify please if we need to support windowSize = 1000 in mainnet? it basically will require each node to use 7gb just to store votes, so we will have to figure out something in order to support that.

This is a question for research, will discuss with @tal-m

File: tortoise/verifying_tortoise.go
Comment (lrettig) [reply tortoise/verifying_tortoise.go]: okay. that's counterintuitive from the design. please document that clearly :) thanks!

**Comment by dshulyak** [comment]
@lrettig @antonlerner i pushed rebased changes. let me know if you want me to change something

**Comment by dshulyak** [comment]
Weird, i have different hare tests hanging now

**Comment by dshulyak** [comment]
lol, why they all passed if executed separately

**Comment by dshulyak** [comment]
actually not hare i was confused by logs. looks like this TestLayerPatterns/heal_after_bad_layers test hangs on CI

**Comment by dshulyak** [comment]
It is even more convoluted than that.. there are two tests in hare that took 14 minutes to execute
```
2021-08-27T08:57:11.5759012Z === RUN   Test_multipleCPs
2021-08-27T09:01:49.2218036Z === RUN   Test_multipleCPsAndIterations
2021-08-27T09:11:06.1888346Z === RUN   Test_HarePreRoundEmptySet
```

File: common/types/block.go
Comment (lrettig) [inline common/types/block.go:22]: Could you please open (and link) an issue to track/discuss this?

File: p2p/service/sim.go
Comment (lrettig) [inline p2p/service/sim.go]: As I commented in #2730, I'm concerned that this will cause issues in app test, let's see

File: tortoise/state.go
Comment (lrettig) [inline tortoise/state.go]: You can remove the comment here and on the two following lines, the context has been lost and I think it's not helpful

File: tortoise/verifying_tortoise.go
Comment (lrettig) [inline tortoise/verifying_tortoise.go]:
```suggestion
		t.GoodBlocksIndex[id] = false // false means good block, not flushed
```

File: tortoise/verifying_tortoise.go
Comment (lrettig) [inline tortoise/verifying_tortoise.go]:
```suggestion
			t.GoodBlocksIndex[b.ID()] = false // false means good block, not flushed
```

**Review by lrettig — CHANGES_REQUESTED** [review]

**Comment by lrettig** [comment]
bors try

File: tortoise/verifying_tortoise.go
Comment (dshulyak) [reply tortoise/verifying_tortoise.go]: i commented where this data structure is defined. i don't think that it helps to add it everywhere where map is used

File: p2p/service/sim.go
Comment (dshulyak) [reply p2p/service/sim.go]: it definitely makes tests less flaky. i already merged that change

**Comment by dshulyak** [comment]
@lrettig i removed some debug logs from a loop https://github.com/spacemeshos/go-spacemesh/pull/2683/commits/4a63c34b9d0b917f3b35384b053d499585eda3a0 . even on my machine it takes nearly 100s to print those logs in some tests. if they are needed we must find a more efficient way to log this information

File: common/types/block.go
Comment (dshulyak) [reply common/types/block.go:22]: https://github.com/spacemeshos/go-spacemesh/issues/2731

File: tortoise/verifying_tortoise.go
Comment (lrettig) [reply tortoise/verifying_tortoise.go]: I disagree. I find this line of code very counterintuitive - that setting the map member to "false" actually indicates that the block is _good_ as opposed to _bad._ I think the explanatory comment helps a lot.

**Comment by lrettig** [comment]
> @lrettig i removed some debug logs from a loop [4a63c34](https://github.com/spacemeshos/go-spacemesh/commit/4a63c34b9d0b917f3b35384b053d499585eda3a0) . even on my machine it takes nearly 100s to print those logs in some tests. if they are needed we must find a more efficient way to log this information

Rather than deleting the debug log lines entirely, could you leave them commented out? It would be helpful for future debugging. Let's continue the discussion about the logging issue in #2732.

File: common/types/block.go
Comment (lrettig) [reply common/types/block.go:22]: thanks!

**Comment by lrettig** [comment]
Take a look at #2734, which is going to cause conflicts here

**Comment by dshulyak** [comment]
> Take a look at #2734, which is going to cause conflicts here

just one more part of the db key

**Comment by dshulyak** [comment]
@lrettig is anything else needs to be changed?

**Comment by dshulyak** [comment]
bors try

**Review by lrettig — APPROVED** [review]
Approving, but just one question... what happens if, theoretically, you were to "restore" a persisted tortoise but with different non-state params, e.g., Hdist, Zdist, ConfidenceInterval, etc.?

**Comment by dshulyak** [comment]
> Approving, but just one question... what happens if, theoretically, you were to "restore" a persisted tortoise but with different non-state params, e.g., Hdist, Zdist, ConfidenceInterval, etc.?

i don't know precisely what is going to happen since it mostly depends on the tortoise logic. but in general, i think we need a general solution to keep consensus critical parameters safe from accidental change. tortoise, tortoise beacon, post, probably hare all require some parameters that can't change arbitrarily after the network was started. i don't think that we want to save them in each module separately.

@lrettig do you know how eth1/eth2 deals with it? i can check in filecoin

**Comment by dshulyak** [comment]
bors merge

**Comment by antonlerner** [comment]
bors try

**Pull request closed without merging**
